climada/engine/uncertainty.py

```diff
@@ -0,0 +1,592 @@
+    def calc_cost_benefit_sobol_sensitivity(self):
+        raise NotImplementedError()
+
+
+    def est_comp_time(self):
+        """
+        Estimate the computation time
+
+        Returns
+        -------
+        None.
+
+        """
+        raise NotImplementedError()
+
+
+    def calc_impact_distribution(self,
+                                 rp=None,
+                                 calc_eai_exp=False,
+                                 calc_at_event=False,
+                                 ):
+        """
+        Computes the impact for each of the parameters set defined in
+        uncertainty.params.
+
+        By default, the aggregated average annual impact
+        (impact.aai_agg) and the excees impact at return periods (rp) is
+        computed and stored in self.aai_freq. Optionally, the impact at
+        each centroid location is computed (this may require a larger
+        amount of memory if the number of centroids is large).
+
+        Parameters
+        ----------
+        rp : list(int), optional
+            Return period in years to be computed.
+            The default is [5, 10, 20, 50, 100, 250].
+        calc_eai_exp : boolean, optional
+            Toggle computation of the impact at each centroid location.
+            The default is False.
+        calc_at_event : boolean, optional
+            Toggle computation of the impact for each event.
+            The default is False.
+
+        Returns
+        -------
+        None.
+
+        """
+
+        if rp is None:
+            rp=[5, 10, 20, 50, 100, 250]
+
+        aai_agg_list = []
+        freq_curve_list = []
+        if calc_eai_exp:
+            eai_exp_list = []
+        if calc_at_event:
+            at_event_list = []
+
+        self.rp = rp
+        self.calc_eai_exp = calc_eai_exp
+        self.calc_at_event = calc_at_event
+
+        if self.pool:
+            chunksize = min(self.n_runs // self.pool.ncpus, 100)
+            impact_metrics = self.pool.map(self._map_impact_eval,
+                                           self.params.iterrows(),
+                                           chunsize = chunksize)
+
+            [aai_agg_list,
+             freq_curve_list,
+             eai_exp_list,
+             at_event_list] = list(zip(*impact_metrics))
+
+        else:
+
+            impact_metrics = map(self._map_impact_eval, self.params.iterrows())
+
+            [aai_agg_list,
+             freq_curve_list,
+             eai_exp_list,
+             at_event_list] = list(zip(*impact_metrics))
+
+        df_aai_freq = pd.DataFrame(freq_curve_list,
+                                   columns=['rp' + str(n) for n in rp])
+        df_aai_freq['aai_agg'] = aai_agg_list
+        self.aai_freq = df_aai_freq
+
+        if calc_eai_exp:
+            df_eai_exp = pd.DataFrame(eai_exp_list)
+            self.eai_exp = df_eai_exp
+
+        if calc_at_event:
+            df_at_event = pd.DataFrame(at_event_list)
+            self.at_event = df_at_event
```

Comment:
at the end of this func we could aggregate all resulting dfs in a dict or something. Now it seems sobol indexes are only calculated based on self.aai_freq (line 427); we could instead iterate first over the dfs and then over the variables in each df.